### PR TITLE
Added example demonstrating how to create and use a wrapping executor

### PR DESCRIPTION
**Pull request by hkaiser**
- this executor wraps any other executor and makes sure that given start
  and stop functions are being called for each HPX thread created by the
  executor

### CODE REVIEW

File: examples/quickstart/executor_with_thread_hooks.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    auto on_stop = [&]() { ++stops; };
+
+    auto exec = executor_example::make_executor_with_thread_hooks(
+        hpx::parallel::execution::par.executor(), on_start, on_stop);
+
+    hpx::parallel::for_loop(hpx::parallel::execution::par.on(exec), 0, v.size(),
+        [](std::size_t i) {});
+
+    std::cout << "Executed " << starts.load() << " starts and " << stops.load()
+              << " stops\n";
```
Comment (msimberg) [inline examples/quickstart/executor_with_thread_hooks.cpp]:
```suggestion
              << " stops\n";
              HPX_ASSERT(starts > 0);
              HPX_ASSERT(stops > 0);
```
just as a sanity check?

Comment (hkaiser) [reply examples/quickstart/executor_with_thread_hooks.cpp]: Ok, I'll add those tests, but will do so a `HPX_TEST_NEQ(starts, std::size_t(0))`.

Comment (hkaiser) [reply examples/quickstart/executor_with_thread_hooks.cpp]: Hold on... This is an example, not a test. So, `HPX_ASSERT` it is.

File: examples/quickstart/executor_with_thread_hooks.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+// The purpose of this example is to show how to rite an executor that wraps
+// any other executor and adds a hook into thread start and thread exit allowing
+// to associate custom thread data with the tasks that are created by the
+// underlying executor.
+
+#include <hpx/hpx_main.hpp>
+
+#include <hpx/algorithms.hpp>
+#include <hpx/execution.hpp>
+#include <hpx/functional.hpp>
```
Comment (msimberg) [inline examples/quickstart/executor_with_thread_hooks.cpp]: Just inspect left on this one: https://173022-4455628-gh.circle-artifacts.com/0/hpx_inspect_report.html. It needs `<cstddef>` and `hpx/assertion.hpp`.

**Pull request merged**